include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+#endif
+
+/**
+ * NOX configuration data
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+  typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData
+                                       additional_data_nox;
+  Teuchos::RCP<Teuchos::ParameterList> parameters_nox =
+    Teuchos::rcp(new Teuchos::ParameterList);
+#endif
+
+/**
+ * PETSc SNES configuration data
+ */
+#ifdef DEAL_II_WITH_PETSC
+  typename PETScWrappers::NonlinearSolverData additional_data_petsc_snes;
+#endif
+
+  /**
+   * Data transfer function
+   */
+  void
+  data_transfer(const AdditionalData &additional_data);
+
+  /**
+   * Solve with PETSc SNES. Internal functions specialized for PETSc
+   * Vectors.
+   */
+  void
+  solve_with_petsc(VectorType &initial_guess_and_solution);
+};
+
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::data_transfer(
+  const AdditionalData &additional_data)
+{
+#ifdef DEAL_II_WITH_SUNDIALS
+  // These if statements pass on the strategy to the other nonlinear solvers
+  if (additional_data.strategy ==
+      NonlinearSolverSelector<VectorType>::AdditionalData::linesearch)
+    additional_data_kinsol.strategy =
+      SUNDIALS::KINSOL<VectorType>::AdditionalData::linesearch;
+  else if (additional_data.strategy ==
+           NonlinearSolverSelector<VectorType>::AdditionalData::newton)
+    additional_data_kinsol.strategy =
+      SUNDIALS::KINSOL<VectorType>::AdditionalData::newton;
+  else if (additional_data.strategy ==
+           NonlinearSolverSelector<VectorType>::AdditionalData::picard)
+    additional_data_kinsol.strategy =
+      SUNDIALS::KINSOL<VectorType>::AdditionalData::picard;
+
+  // Setting data points in the KINSOL class from the NonlinearSolverSelector
+  // class
+  additional_data_kinsol.maximum_non_linear_iterations =
+    additional_data.maximum_non_linear_iterations;
+  additional_data_kinsol.function_tolerance =
+    additional_data.function_tolerance;
+  additional_data_kinsol.step_tolerance = additional_data.step_tolerance;
+  additional_data_kinsol.anderson_subspace_size =
+    additional_data.anderson_subspace_size;
+#endif
+
+// Do the same thing we did above but with NOX
+#ifdef DEAL_II_WITH_TRILINOS
+  parameters_nox->set("Nonlinear Solver", "Line Search Based");
+  Teuchos::ParameterList &Line_Search = parameters_nox->sublist("Line Search");
+  Line_Search.set("Method", "Full Step");
+
+  additional_data_nox.max_iter = additional_data.maximum_non_linear_iterations;
+  additional_data_nox.abs_tol  = additional_data.function_tolerance;
+  additional_data_nox.rel_tol  = additional_data.relative_tolerance;
+#endif
+
+#ifdef DEAL_II_WITH_PETSC
+  additional_data_petsc_snes.options_prefix = "";
+
+  if (additional_data.anderson_subspace_size > 0 &&
+      additional_data.strategy ==
+        NonlinearSolverSelector<VectorType>::AdditionalData::picard)
+    {
+      additional_data_petsc_snes.snes_type = "anderson";
+      // TODO additional_data.anderson_subspace_size;
```

Comment:
might be a good time to address this right now :-)

Reply:
@stefanozampini Do you know how I can set ``-snes_anderson_m X`` programmatically for the SNES solver?

Reply:
It appears there is no API currently, only command line options. You can use `PetscOptionsSetValue` https://petsc.org/main/manualpages/Sys/PetscOptionsSetValue for now. I will add the API that will appear in the new release of PETSc. We can put a general call `snes_anderson_set_m`  in the compatibility header

Reply:
Does ``PETScWrappers::set_option_value("-snes_anderson_m","15")`` work or do I have to worry about the options prefix that is set?

Reply:
You should  call SNESGetOptionsPrefix and append snes_anderson_m